client/src/__tests__/TokenForm.test.js

```diff
@@ -0,0 +1,399 @@
+import React from "react";
+import {
+  fireEvent,
+  waitForElement,
+  render,
+  wait,
+} from "@testing-library/react";
+import { act } from "react-dom/test-utils";
+
+import { TokenForm } from "../components";
+
+describe("Token Form Component", () => {
+  it("renders the page", () => {
+    const wrapper = render(<TokenForm />);
+    expect(wrapper).toBeDefined();
+  });
+
+  it("will not continue if token type and token format is not selected", async () => {
+    const { getByText, container } = render(<TokenForm />);
+
+    const button = await waitForElement(() => container.querySelector("Form"));
+
+    await wait(() => {
+      fireEvent.submit(button);
+    });
+
+    expect(getByText("Must select a token type")).not.toBeNull();
+    expect(getByText("Must select a token format")).not.toBeNull();
+  });
+
+  it("renders scopes step if token type and token format has been entered and token type is OAuth", async () => {
+    const { getByText, container, getByLabelText } = render(<TokenForm />);
+    const TypeInput = await waitForElement(() => getByLabelText("OAuth"));
+    const FormatInput = await waitForElement(() => getByLabelText("Bare"));
+    const button = await waitForElement(() => container.querySelector("Form"));
+
+    await wait(() => {
+      fireEvent.click(TypeInput);
+    });
+
+    await wait(() => {
+      fireEvent.click(FormatInput);
+    });
+
+    await wait(() => {
+      fireEvent.submit(button);
+    });
+
+    expect(getByText("Enter scopes")).not.toBeNull();
+  });
+
+  it("renders enter audience step if token type and token format has been entered and token type is JWT", async () => {
+    const { getByText, container, getByLabelText } = render(<TokenForm />);
+    const TypeInput = await waitForElement(() => getByLabelText("JWT"));
+    const FormatInput = await waitForElement(() => getByLabelText("Bare"));
+    const button = await waitForElement(() => container.querySelector("Form"));
+
+    await wait(() => {
+      fireEvent.click(TypeInput);
+    });
+
+    await wait(() => {
+      fireEvent.click(FormatInput);
+    });
+
+    await wait(() => {
+      fireEvent.submit(button);
+    });
+
+    expect(getByText("Enter audience")).not.toBeNull();
+  });
+
+  it("will not continue if scopes has not been entered when token type is OAuth", async () => {
+    const { container, getByLabelText, getByText } = render(<TokenForm />);
+    const TypeInput = await waitForElement(() => getByLabelText("OAuth"));
+    const FormatInput = await waitForElement(() => getByLabelText("Bare"));
+    const button = await waitForElement(() => container.querySelector("Form"));
+
+    await wait(() => {
+      fireEvent.click(TypeInput);
+    });
+
+    await wait(() => {
+      fireEvent.click(FormatInput);
+    });
+
+    await wait(() => {
+      fireEvent.submit(button);
+    });
+
+    await wait(() => {
+      fireEvent.submit(button);
+    });
```

Comment:
How come this section is repeated?

Reply:
Once for the token type and format, then the second to check if it does not move forward for the scopes/audience step if scopes/audience has not been selected.

Reply:
Ah okay thanks for explaining!

Reply:
Can we add a comment here explaining this? Because it can potentially be confusing for future devs on the preoject.